README.md
![](docs/img/fuzz_pint_logo.jpg)
# FuzzyPint

FuzzyPint is a "wrapper" around [Pint](https://pint.readthedocs.io) objects that adds +/- error margins (which can be asymmetric) for the purpose of propagating "significant digits" through calculations.

## TL;DR
- Create a bunch of values with **units** and **error margin**.
- Do a bunch of math on them.
- Get a result with a (propagated) error margin.
- Display the result, with units, with the number of **significant digits** reflected by that result's error margin.

## Overview
FuzzyPint values correctness and simplicity over performance. Every math operation on (a pair of) FuzzyPint objects performs 5 calculations: A nominal calculation on the nominal magnitudes, and all 4 possible combinations of worst-case min/max error propagations.  FuzzyPint does not try to be "smart" about how errors propagate; instead it brute-forces all calculations to see what the resulting worst-case min/max error would be, then includes that (worst case) error margin in the result.


## Installation
There is no pip release (yet). FuzzyPint is a single file with a single external dependency (the [`pint`](https://pint.readthedocs.io) package), so for now you can just download `fuzzy_pint.py` to use it in a project.

To use FuzzyPint, import the FuzzyPint object like this:
```python
from fuzzy_pint import FuzzyPint
```

## Example
### Basic Calculations
Let's say you measure a resistor in the lab as "1000.2345 Ohms", and you know that the error of your multimeter (for resistance measurements) is +/- 0.01 ohms. You would enter that value like this:

```
>>> from fuzzy_pint import FuzzyPint
>>> r = FuzzyPint(1000.2345, "ohm", 0.01, -0.01)
>>> r
<FuzzyPint(1000.23, "ohm", err_p=0.01, err_n=-0.01)>
>>>
```

Now let's say you measure the voltage across that resistor as 5.4321 V, and you know that the error of your multimeter (for voltage measurements) is +/- 0.001 V. You would enter that value like this:

```
>>> v=FuzzyPint(5.4321, "volt", 0.001, -0.001)
>>> v
<FuzzyPint(5.4321, "volt", err_p=0.001, err_n=-0.001)>
>>>
```

You could then calculate the power dissipation in the resistor (P = V²/R) like this:

```
>>> p = v**2 / r
>>> p
<FuzzyPint(0.0295008, "volt ** 2 / ohm", err_p=1.11577e-05, err_n=-1.11555e-05)>
>>> 
```

Notice that FuzzyPint has calculated the resulting (propagated) error margin. The units are in "volt² / ohms" so let's convert that to milliwatts...

```
>>> p
<FuzzyPint(29.5008, "milliwatt", err_p=0.0111577, err_n=-0.0111555)>
>>>
```

[Pint](https://pint.readthedocs.io) does the heavy lifting here (units conversion, including SI prefix scaling of the nominal magnitude), and FuzzyPint scales the errors to match the new (scaled) magnitude.

Now we can render the final result showing only the "significant" digits like this...

```
>>> p.significant()
'29.50 mW'
>>>
```
### JSON Serialization
If we wish to store this result in a JSON file for later use we can serialize the FuzzyPint object like this:

```
>>> d = p.to_serializable()
>>> d
{'value': 29.50079, 'units': 'milliwatt', 'err_p': 0.011157703020417409, 'err_n': -0.011155480409783675}
>>>
```

The `.to_serializable()` method returns a (JSON Serializable) dict:

```
>>> import json
>>> j = json.dumps(d, indent=4)
>>> print(j)
{
    "value": 29.50079,
    "units": "milliwatt",
    "err_p": 0.011157703020417409,
    "err_n": -0.011155480409783675
}
>>>
```

And we can recover a serialized FuzzyPint object like this:
```
>>> v_recovered = FuzzyPint.from_serializable(d)
>>> v_recovered
<FuzzyPint(29.5008, "milliwatt", err_p=0.0111577, err_n=-0.0111555)>
>>>
```

## Demo
You can demo the functionality by executing the module:

```
$ ./fuzzy_pint.py
fuzzy_pint 0.0.2
FuzzyPint.__repr__():
    v1: <FuzzyPint(2.73, "volt", err_p=0.13, err_n=-0.13)>
FuzzyPint.__str__():
    v1: 2.73 volt [+0.13, -0.13]
FuzzyPint.pretty():
    v1: 2.73 V [+0.13, -0.13]
FuzzyPint.significant():
    v1: 2.7 V
Dimensionless:
    d1: <FuzzyPint(123.4, "dimensionless", err_p=0, err_n=0)>
Conversion:
    w1=<FuzzyPint(1.23, "ampere * volt", err_p=0.01, err_n=-0.01)>
    w1.to("watt")=<FuzzyPint(1.23, "watt", err_p=0.01, err_n=-0.01)>
Add:
    v1=<FuzzyPint(2.73, "volt", err_p=0.13, err_n=-0.13)>
    v2=<FuzzyPint(9.77, "volt", err_p=0.13, err_n=-0.13)>
    v1+v2=<FuzzyPint(12.5, "volt", err_p=0.26, err_n=-0.26)>
    (v1+v2).pretty()='12.5 V [+0.26, -0.26]'
    (v1+v2).significant()='12.5 V'
Subtract:
    v1=<FuzzyPint(2.73, "volt", err_p=0.13, err_n=-0.13)>
    v2=<FuzzyPint(9.77, "volt", err_p=0.13, err_n=-0.13)>
    v1-v2=<FuzzyPint(-7.04, "volt", err_p=0.26, err_n=-0.26)>
    (v1-v2).pretty()='-7.04 V [+0.26, -0.26]'
    (v1-v2).significant()='7.0 V'
Multiply:
    v1=<FuzzyPint(2.73, "volt", err_p=0.1, err_n=-0.2)>
    i1=<FuzzyPint(21.97, "ampere", err_p=0.3, err_n=-0.4)>
    v1*i1=<FuzzyPint(59.9781, "ampere * volt", err_p=3.046, err_n=-5.406)>
    (v1*i1).pretty()='59.9781 A·V [+3.046, -5.406]'
    (v1*i1).significant()='60 A·V'
Divide:
    v1=<FuzzyPint(12.73, "volt", err_p=0.1, err_n=-0.2)>
    r1=<FuzzyPint(1234, "ohm", err_p=50, err_n=-50)>
    v1/r1=<FuzzyPint(0.010316, "volt / ohm", err_p=0.000520103, err_n=-0.000557478)>
    (v1/r1).pretty()='0.010316 V/Ω [+0.000520103, -0.000557478]'
    (v1/r1).significant()='0.0103 V/Ω'
Exponent:
    v1=<FuzzyPint(12.73, "volt", err_p=0.1, err_n=-0.2)>
    v1**2=<FuzzyPint(162.053, "volt ** 2", err_p=2.556, err_n=-5.052)>
    (v1**2).pretty()='162.053 V² [+2.556, -5.052]'
    (v1**2).significant()='162 V²'
Dimensionless Multiply:
    v1=<FuzzyPint(2.73, "volt", err_p=0.1, err_n=-0.2)>
    i1=2.0
    v1*i1=<FuzzyPint(5.46, "volt", err_p=0.2, err_n=-0.4)>
    (v1*i1).pretty()='5.46 V [+0.2, -0.4]'
    (v1*i1).significant()='5.5 V'
Prefix Scale Conversion: Example: 1.23456 from V -> mV:
    v1=<FuzzyPint(1.23456, "volt", err_p=0.0001, err_n=-0.0001)>
    v1*1000=<FuzzyPint(1234.56, "volt", err_p=0.1, err_n=-0.1)>
    (v1).to("millivolt")=<FuzzyPint(1234.56, "millivolt", err_p=0.1, err_n=-0.1)>
    (v1).to("millivolt").pretty()='1234.56 mV [+0.1, -0.1]'
    (v1).to("millivolt").significant()='1234.6 mV'
Significance Precedence with Asymmetric Error:
    v1=<FuzzyPint(1234.57, "volt", err_p=0.1, err_n=-0.1)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.1, err_n=-0.01)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.1, err_n=-0.001)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.01, err_n=-0.1)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.001, err_n=-0.1)>
    v1.significant()='1234.6 V'

Significance:
    v1=<FuzzyPint(1234.57, "volt", err_p=2000, err_n=-2000)>
    v1.significant()='1000 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=200, err_n=-200)>
    v1.significant()='1200 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=20, err_n=-0.2)>
    v1.significant()='1230 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=2, err_n=-2)>
    v1.significant()='1235 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.1, err_n=-0.1)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.12, err_n=-0.12)>
    v1.significant()='1234.6 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.01, err_n=-0.01)>
    v1.significant()='1234.57 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.009, err_n=-0.009)>
    v1.significant()='1234.568 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=0.0009, err_n=-0.0009)>
    v1.significant()='1234.5678 V'

    v1=<FuzzyPint(1234.57, "volt", err_p=9e-05, err_n=-9e-05)>
    v1.significant()='1234.56780 V'

Significance, Zero Padding:
    v1=<FuzzyPint(1, "volt", err_p=0.1, err_n=-0.1)>
    v1.significant()='1.0 V'

    v1=<FuzzyPint(1, "volt", err_p=0.01, err_n=-0.01)>
    v1.significant()='1.00 V'

    v1=<FuzzyPint(1, "volt", err_p=0.001, err_n=-0.001)>
    v1.significant()='1.000 V'

JSON serialization
    v1: <FuzzyPint(2.55121, "volt", err_p=0.13, err_n=-0.13)>
    v1: 2.551212121212 volt [+0.13, -0.13]
    v1.significant()='2.6 V'
    v1.to_serializable()={'value': 2.5512, 'units': 'volt', 'err_p': 0.13, 'err_n': -0.13}
    json.dumps(v1.to_serializable())='{"value": 2.5512, "units": "volt", "err_p": 0.13, "err_n": -0.13}'
    v1_serialized={'value': 2.5512, 'units': 'volt', 'err_p': 0.13, 'err_n': -0.13}
    FuzzyPint.from_serializable(v1_serialized)=<FuzzyPint(2.5512, "volt", err_p=0.13, err_n=-0.13)>
$ 
```

## Why not use Pint's `Measurement` object?
Pint has a native `Measurement` object which can represent measurement error in a manner similar to FuzzyPint, but it has three significant limitations:

- It does not propagate errors properly for non-linear operations
    - This is a known issue. Pint's [Measurement](https://pint.readthedocs.io/en/stable/advanced/measurement.html) documentation says "Only linear combinations are currently supported."
- It does not support asymmetric error margins (e.g. `10 V +1/-2`)
- It does not natively support rendering a value's significant figures based on that value's error margin.

### An example of the non-linear issue with Pint's `Measurement` object

If we square a voltage of 10 V +/- 1 V using Pint's native `Measurement` object we get:

```
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> v = (10 * ureg.volt).plus_minus(1)
>>> print(v)
(10.0 +/- 1.0) volt
>>> print(v**2)
(100 +/- 20) volt ** 2
>>>
```

Pint's native `Measurement` object incorrectly reports the resulting error as `+/- 20`, but the error should be `+21 / -19` (because the worst case positive error would occur at `(10+1) * (10+1) == 121 == 100 + 21` and the worst case negative error would occur at `(10-1) * (10-1) == 81 == 100 - 19`).

For the same calculation FuzzyPint correctly reports the expected error of `+21 / -19`:

```
>>> from fuzzy_pint import FuzzyPint
>>> v = FuzzyPint(10, "volt", 1, -1)
>>> print(v)
10 volt [+1, -1]
>>> print(v**2)
100 volt ** 2 [+21.0, -19.0]
>>>
```


## Development Status
As of version `0.0.2`:

- Supported operations:
    - Add
    - Subtract
    - Multiply
    - Divide
    - Power (i.e. Exponentiation)
- There are no tests (yet); just a demonstration of functionality.
- There are no (known) bugs.
- I am plumbing this into a "real" data analysis project.  Once that project has successfully "kicked the tires" in a real-world application I will consider the API "stable" and I will add proper test cases.

## Shout out

[Pint](https://pint.readthedocs.io) is fantastic and I love it immensely! The ability to carry units through calculations significantly guards against human error when combining or scaling physical quantities.  If you are manipulating physical quantities in Python then you should give Pint a try.

CHANGELOG.md
# FuzzyPint Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
*(None)*

## v0.0.2 - 2023-06-17
### Added
- Initial Release.

fuzzy_pint.py
#!/usr/bin/env python3

# Standard Library
from math import log10, floor
from typing import Tuple
import json

# Library
import pint


# Local
# from util_pint import pint_dict_to_pint, pint_to_pint_dict

APP_NAME = 'fuzzy_pint'
APP_VERSION = '0.0.2'
DEBUG_ENABLE = False

UREG = pint.UnitRegistry()


def main():
    """Fuzzy Pint demo."""
    indent = ' ' * 4
    print(f'{APP_NAME} {APP_VERSION}')

    v1 = FuzzyPint(2.73, 'volt', 0.13, -0.13)
    v2 = FuzzyPint(9.77, 'volt', 0.13, -0.13)
    i1 = FuzzyPint(21.97, 'ampere', 0.21, -0.21)
    w1 = FuzzyPint(1.23, 'ampere*volt', 0.01, -0.01)
    d1 = FuzzyPint(123.4)

    print('FuzzyPint.__repr__():')
    print(f'{indent}v1: {v1!r}')
    print('FuzzyPint.__str__():')
    print(f'{indent}v1: {v1}')
    print('FuzzyPint.pretty():')
    print(f'{indent}v1: {v1.pretty()}')
    print('FuzzyPint.significant():')
    print(f'{indent}v1: {v1.significant()}')
    print('Dimensionless:')
    print(f'{indent}d1: {d1!r}')
    print('Conversion:')
    print(f'{indent}{w1=}')
    print(f'{indent}{w1.to("watt")=}')

    print('Add:')
    print(f'{indent}{v1=}')
    print(f'{indent}{v2=}')
    print(f'{indent}{v1+v2=}')
    print(f'{indent}{(v1+v2).pretty()=}')
    print(f'{indent}{(v1+v2).significant()=}')

    print('Subtract:')
    print(f'{indent}{v1=}')
    print(f'{indent}{v2=}')
    print(f'{indent}{v1-v2=}')
    print(f'{indent}{(v1-v2).pretty()=}')
    print(f'{indent}{(v1-v2).significant()=}')

    print('Multiply:')
    v1 = FuzzyPint(2.73, 'volt', 0.1, -0.2)
    i1 = FuzzyPint(21.97, 'ampere', 0.3, -0.4)
    print(f'{indent}{v1=}')
    print(f'{indent}{i1=}')
    print(f'{indent}{v1*i1=}')
    print(f'{indent}{(v1*i1).pretty()=}')
    print(f'{indent}{(v1*i1).significant()=}')

    print('Divide:')
    v1 = FuzzyPint(12.73, 'volt', 0.1, -0.2)
    r1 = FuzzyPint(1234, 'ohm', 50, -50)
    print(f'{indent}{v1=}')
    print(f'{indent}{r1=}')
    print(f'{indent}{v1/r1=}')
    print(f'{indent}{(v1/r1).pretty()=}')
    print(f'{indent}{(v1/r1).significant()=}')

    print('Exponent:')
    v1 = FuzzyPint(12.73, 'volt', 0.1, -0.2)
    print(f'{indent}{v1=}')
    print(f'{indent}{v1**2=}')
    print(f'{indent}{(v1**2).pretty()=}')
    print(f'{indent}{(v1**2).significant()=}')

    print('Dimensionless Multiply:')
    v1 = FuzzyPint(2.73, 'volt', 0.1, -0.2)
    i1 = 2.0
    print(f'{indent}{v1=}')
    print(f'{indent}{i1=}')
    print(f'{indent}{v1*i1=}')
    print(f'{indent}{(v1*i1).pretty()=}')
    print(f'{indent}{(v1*i1).significant()=}')

    print('Prefix Scale Conversion: Example: 1.23456 from V -> mV:')
    v1 = FuzzyPint(1.23456, 'volt', 0.0001, -0.0001)
    print(f'{indent}{v1=}')
    print(f'{indent}{v1*1000=}')
    print(f'{indent}{(v1).to("millivolt")=}')
    print(f'{indent}{(v1).to("millivolt").pretty()=}')
    print(f'{indent}{(v1).to("millivolt").significant()=}')

    # print('Significance:')
    for v1 in (
        'Significance Precedence with Asymmetric Error:',
        FuzzyPint(1234.5678, 'volt', 0.1, -0.1),
        FuzzyPint(1234.5678, 'volt', 0.1, -0.01),
        FuzzyPint(1234.5678, 'volt', 0.1, -0.001),
        FuzzyPint(1234.5678, 'volt', 0.01, -0.1),
        FuzzyPint(1234.5678, 'volt', 0.001, -0.1),
        'Significance:',
        FuzzyPint(1234.5678, 'volt', 2000.0, -2000.0),
        FuzzyPint(1234.5678, 'volt', 200.0, -200.0),
        FuzzyPint(1234.5678, 'volt', 20.0, -0.2),
        FuzzyPint(1234.5678, 'volt', 2.0, -2.0),
        FuzzyPint(1234.5678, 'volt', 0.1, -0.1),
        FuzzyPint(1234.5678, 'volt', 0.12, -0.12),
        FuzzyPint(1234.5678, 'volt', 0.01, -0.01),
        FuzzyPint(1234.5678, 'volt', 0.009, -0.009),
        FuzzyPint(1234.5678, 'volt', 0.0009, -0.0009),
        FuzzyPint(1234.5678, 'volt', 0.00009, -0.00009),
        'Significance, Zero Padding:',
        FuzzyPint(1.0, 'volt', 0.1, -0.1),
        FuzzyPint(1.0, 'volt', 0.01, -0.01),
        FuzzyPint(1.0, 'volt', 0.001, -0.001),
    ):
        if isinstance(v1, str):
            # This is a heading
            print(v1)
            continue
        print(f'{indent}{v1=}')
        print(f'{indent}{v1.significant()=}')
        print()

    print('JSON serialization')
    v1 = FuzzyPint(2.551212121212, 'volt', 0.13, -0.13)
    print(f'{indent}v1: {v1!r}')
    print(f'{indent}v1: {v1}')
    print(f'{indent}{v1.significant()=}')
    print(f'{indent}{v1.to_serializable()=}')
    print(f'{indent}{json.dumps(v1.to_serializable())=}')
    v1_serialized = v1.to_serializable()
    print(f'{indent}{v1_serialized=}')
    print(f'{indent}{FuzzyPint.from_serializable(v1_serialized)=}')

class FuzzyPint:
    def __init__(self, magnitude: float, units: str = None, err_p: float = 0.0, err_n: float = 0.0):
        assert err_p >= 0
        assert err_n <= 0
        units = 'dimensionless' if units is None else units
        units = UREG(units) if isinstance(units, str) else units
        self._quantity = magnitude * units
        self._err_p = err_p
        self._err_n = err_n

    def clone(self):
        return FuzzyPint(self._quantity.m, self._quantity.units, self._err_p, self._err_n)

    def to(self, units: str):
        """Convert units.

        Some conversions (such as 'volts' to 'millivolts') may change the scale of the property's
        magnitude, in which case it is necessary for us to similarly scale the associated error.

        This conversion auto-detects such scale changes and scales the errors accordingly.
        """
        if isinstance(units, str):
            units = UREG(units)
        new_object = self.clone()
        new_object._quantity = self._quantity.to(units)
        if new_object._quantity.m != self._quantity.m:
            # The units change scaled the magnitude, so scale the errors similarly.
            scale = new_object._quantity.m / self._quantity.m
            new_object._err_p *= scale
            new_object._err_n *= scale
        return new_object
    
    def to_serializable(self, extended_precision=3):
        q_magnitude, _ = self.significant_magnitude(extended_precision=extended_precision)
        return {
            'value': q_magnitude,
            'units': f'{self._quantity.units:D}',
            'err_p': self._err_p,
            'err_n': self._err_n,
        }
    
    @classmethod
    def from_serializable(cls, object: dict) -> 'FuzzyPint':
        return FuzzyPint(
            object['value'],
            object['units'],
            object['err_p'],
            object['err_n'],
        )

    def __add__(self, b):
        return FuzzyPint._apply_function(self, b, FuzzyPint._add)

    def __sub__(self, b):
        return FuzzyPint._apply_function(self, b, FuzzyPint._sub)

    def __mul__(self, b):
        return FuzzyPint._apply_function(self, b, FuzzyPint._multiply)

    def __truediv__(self, b):
        return FuzzyPint._apply_function(self, b, FuzzyPint._divide)

    def __pow__(self, b):
        return FuzzyPint._apply_function(self, b, FuzzyPint._pow)

    @staticmethod
    def _apply_function(a: 'FuzzyPint', b: 'FuzzyPint', function):
        # Convert dimensionless quantities to FuzzyPint objects if supplied
        if isinstance(a, (float, int)):
            a = FuzzyPint(a)
        if isinstance(b, (float, int)):
            b = FuzzyPint(b)

        quantity = function(a._quantity, b._quantity)
        err_p, err_n = FuzzyPint._get_error(a, b, function)
        return FuzzyPint(quantity.m, quantity.units, err_p, err_n)

    @staticmethod
    def _get_error(a: 'FuzzyPint', b: 'FuzzyPint', function):
        a_m = a._quantity.m
        b_m = b._quantity.m
        nominal_m = function(a_m, b_m)

        errors = []
        for a_m_speculative, b_m_speculative in (
            (a_m + a._err_p, b_m + b._err_p),
            (a_m + a._err_n, b_m + b._err_p),
            (a_m + a._err_p, b_m + b._err_n),
            (a_m + a._err_n, b_m + b._err_n),
        ):
            result = function(a_m_speculative, b_m_speculative)
            error = result - nominal_m
            errors.append(error)
        err_p = max(errors)
        err_n = min(errors)
        assert err_p >= 0
        assert err_n <= 0

        return err_p, err_n

    @staticmethod
    def _add(a, b):
        return a + b

    @staticmethod
    def _sub(a, b):
        return a - b

    @staticmethod
    def _multiply(a, b):
        return a * b

    @staticmethod
    def _divide(a, b):
        return a / b

    @staticmethod
    def _pow(a, b):
        return a ** b

    def pretty(self):
        return f'{self._quantity:g~P} [+{self._err_p:g}, {self._err_n:g}]'

    def significant(self):

        q_magnitude, decimal_rounding_digits = self.significant_magnitude()
        quantity = q_magnitude * self._quantity.units

        return f'{quantity:0.{decimal_rounding_digits}f~P}'
    
    def significant_magnitude(self, extended_precision: int = 0) -> Tuple[float, int]:
        """Return the significant part of the magnitude.

        The caller may request additional precision. This is typically used when
        storing values that will be used for future calculations (so that the
        precision is not prematurely truncated).

        Args:
            extended_precision (int, optional): Additional decimal digits of precision
            to use. Defaults to 0.

        Returns:
            Tuple[float, int]:
                ( The quantity (rounded to significant digits),
                  The number of decimal digits of precision )
        """
        q_magnitude = self._quantity.m
        q_significand, q_exponent, q_is_negative = self._float_to_scientific(q_magnitude)
        _debug_print(f'{q_significand=}, {q_exponent=}, {q_is_negative=}')

        err_p_significand, err_p_exponent, err_p_is_negative = self._float_to_scientific(
            self._err_p / (10 ** extended_precision)
        )
        _debug_print(f'{err_p_significand=}, {err_p_exponent=}, {err_p_is_negative=}')

        err_n_significand, err_n_exponent, err_n_is_negative = self._float_to_scientific(
            self._err_n / (10 ** extended_precision)
        )
        _debug_print(f'{err_n_significand=}, {err_n_exponent=}, {err_n_is_negative=}')

        err_exponent_max = max(err_p_exponent, err_n_exponent)
        _debug_print(f'{err_exponent_max=}')
        # print(f'🟠 {err_exponent_max=}')

        # Strip insignificant digits
        shift_exponent = q_exponent - err_exponent_max
        _debug_print(f'{shift_exponent=}')
        q_significand = round(q_significand * 10 ** (shift_exponent), 0) * 10 ** (-shift_exponent)
        _debug_print(f'{q_significand=}')
        q_magnitude = self._scientific_to_float(q_significand, q_exponent, q_is_negative)
        _debug_print(f'{q_magnitude=}')

        # Round any remaining insignificant "display digits"
        # (because we are working in floating point, there may be a very small epsilon in the
        # floating point representation of q_magnitude, so we will strip it)
        decimal_rounding_digits = abs(err_exponent_max) if err_exponent_max < 0 else 0

        _debug_print(f'{decimal_rounding_digits=}')
        # print(f'🟠 {decimal_rounding_digits=}')
        q_magnitude = round(q_magnitude, decimal_rounding_digits)
        _debug_print(f'{q_magnitude=}')

        return q_magnitude, decimal_rounding_digits

    @staticmethod
    def _float_to_scientific(value: float):
        is_negative = value < 0
        exponent = floor(log10(abs(value)))
        significand = value * 10 ** (-exponent)
        return significand, exponent, is_negative

    @staticmethod
    def _scientific_to_float(significand: float, exponent: int, is_negative: bool) -> float:
        value = significand * 10 ** exponent
        return -value if is_negative else value

    def __repr__(self):
        return (
            f'<FuzzyPint({self._quantity.m:g}, "{self._quantity.units}", err_p={self._err_p:g}, err_n={self._err_n:g})>'
        )

    def __str__(self):
        return f'{self._quantity} [+{self._err_p}, {self._err_n}]'


def _debug_print(text):
    if not DEBUG_ENABLE:
        return
    print(f'🟣  {text}')


if __name__ == "__main__":
    main()
